### ISSUE TITLE
Sorting spherical Voronoi vertices leads to crash in area calculation

### ISSUE THREAD

**Issue by mueslo** [scipy/scipy#13412]
There appears to be a bug in the spherical Voronoi code. If you sort the vertices (e.g. for plotting) then the area calculation fails to work.

#### Reproducing code example:

```
import numpy as np
from scipy.spatial import SphericalVoronoi

points = np.array([[0, 0, 1], [0, 0, -1], [1, 0, 0], [0, 1, 0], [0, -1, 0], [-1, 0, 0], 
                   [1,1,0], [1,-1,0], [-1,1,0], [-1,-1,0], [1,0,1], [1,0,-1], [-1,0,1], [-1,0,-1], [0,1,1], [0,1,-1], [0,-1,1],[0,-1,-1],
                   [1,1,1], [1,1,-1], [1,-1,1], [1,-1,-1], [-1,1,1], [-1,1,-1], [-1,-1,1], [-1,-1,-1]], dtype=float)

points /= np.sqrt((points**2).sum(axis=1))[:, np.newaxis]

print(points)

radius = 1
center = np.array([0, 0, 0])
sv = SphericalVoronoi(points, radius, center)
sv.sort_vertices_of_regions() #<-- if this is not run it works
areas = sv.calculate_areas()
print(areas)
```

#### Error message:

```
ValueError                                Traceback (most recent call last)
<ipython-input-4-5ff09bf5b06f> in <module>
      9 sv = SphericalVoronoi(points, radius, center)
     10 sv.sort_vertices_of_regions()
---> 11 areas = sv.calculate_areas()
     12 print(areas)

~\AppData\Local\Programs\Python\Python38\lib\site-packages\scipy\spatial\_spherical_voronoi.py in calculate_areas(self)
    341             return self._calculate_areas_2d()
    342         elif self._dim == 3:
--> 343             return self._calculate_areas_3d()
    344         else:
    345             raise TypeError("Only supported for 2D and 3D point sets")

~\AppData\Local\Programs\Python\Python38\lib\site-packages\scipy\spatial\_spherical_voronoi.py in _calculate_areas_3d(self)
    265 
    266     def _calculate_areas_3d(self):
--> 267         self.sort_vertices_of_regions()
    268         sizes = [len(region) for region in self.regions]
    269         csizes = np.cumsum(sizes)

~\AppData\Local\Programs\Python\Python38\lib\site-packages\scipy\spatial\_spherical_voronoi.py in sort_vertices_of_regions(self)
    262         if self._dim != 3:
    263             raise TypeError("Only supported for three-dimensional point sets")
--> 264         _voronoi.sort_vertices_of_regions(self._simplices, self.regions)
    265 
    266     def _calculate_areas_3d(self):

_voronoi.pyx in scipy.spatial._voronoi.sort_vertices_of_regions()

ValueError: Buffer dtype mismatch, expected 'npy_intp' but got 'long'
```

#### Scipy/Numpy/Python version information:

```
<<< import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
>>> 1.6.0 1.18.4 sys.version_info(major=3, minor=8, micro=2, releaselevel='final', serial=0)
```

**Comment by pmla** [comment]
Thanks for reporting this @mueslo.

@tylerjereddy I can't reproduce the failure on my machine. It is probably an architecture specific typing problem.

**Comment by mueslo** [comment]
Ah, from the traceback I see simply calling sv.sort_vertices_of_regions() twice is enough to trigger the error, although it should be an idempotent function and run fine. I also tried with the latest numpy (1.19.4), same error

**Comment by mueslo** [comment]
Some additional info: regions type changes


```
import numpy as np
from scipy.spatial import SphericalVoronoi

points = np.array([[0, 0, 1], [0, 0, -1], [1, 0, 0], [0, 1, 0], [0, -1, 0], [-1, 0, 0], 
                   [1,1,0], [1,-1,0], [-1,1,0], [-1,-1,0], [1,0,1], [1,0,-1], [-1,0,1], [-1,0,-1], [0,1,1], [0,1,-1], [0,-1,1],[0,-1,-1],
                   [1,1,1], [1,1,-1], [1,-1,1], [1,-1,-1], [-1,1,1], [-1,1,-1], [-1,-1,1], [-1,-1,-1]], dtype=float)

points /= np.sqrt((points**2).sum(axis=1))[:, np.newaxis]

radius = 1.0
center = np.array([0, 0, 0], dtype=float)
sv = SphericalVoronoi(points, radius, center)
print(sv._simplices.dtype, sv.vertices.dtype, type(sv.regions[0][0]))
sv.sort_vertices_of_regions()
print(sv._simplices.dtype, sv.vertices.dtype, type(sv.regions[0][0]))
sv.sort_vertices_of_regions()
```

int32 float64 <class 'numpy.int64'>
int32 float64 <class 'int'>
<exception traceback>

**Comment by pmla** [comment]
@mueslo Great debugging work. I have pushed a fix in #13414.
